packages/postgres-database/src/simcore_postgres_database/migration/versions/ef817566287e_new_services_vendor_secrets_table.py

```diff
@@ -0,0 +1,42 @@
+"""new services_vendor_secrets table
+
+Revision ID: ef817566287e
+Revises: 0c084cb1091c
+Create Date: 2023-05-18 11:38:51.671239+00:00
+
+"""
+import sqlalchemy as sa
+from alembic import op
+from sqlalchemy.dialects import postgresql
+
+# revision identifiers, used by Alembic.
+revision = "ef817566287e"
+down_revision = "0c084cb1091c"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.create_table(
+        "services_vendor_secrets",
+        sa.Column("service_key", sa.String(), nullable=False),
```

Comment:
no relation to services_meta_data? why?

Reply:
In short: this is not possible with the current design. we would have to normalize the services_metadata table.
More details:

- `services_vendor_secrets` has only `key` column but cannot use a foreign key to `services_metadata` because the latter does not have a unique `key`. Note that the later is unique on (key, version) pair. Therefore there is not a unique correspondence. 
- A way to achieve this is normalizing the current design, i.e. create separately a table for service keys, then another for service versions and then use service_metadata as a mapping table. 
```mermaid
classDiagram
    class services {
        key
    }

    class versions {
        tag
    }

    class services_metadata {
        service_key
        service_version
        description
    }

    services_metadata "1" -- "1" services : key
    services_metadata "1" -- "1" versions : tag
```
NOTE: perhaps the versions table is not necessary .

Then we could use a foreign key from ``services_vendor_secrets` to the new `services` table
```mermaid
classDiagram
    class services {
        key
    }

    class versions {
        tag
    }

    class services_metadata {
        service_key
        service_version
        description
    }
    
    class services_vendor_secrets {
        service_key
        vendor_secrets
    }

    services_metadata "1" -- "1" services : key
    services_metadata "1" -- "1" versions : tag
    services_vendor_secrets "1" -- "1" services : key

```

Reply:
well that is interesting, I did not know about that naming and it makes sense. So if I get your problem is that you cannot create a relation to the key because these are not unique in the services_metadata. I think we do not need the versions table until we need it. Then comes the second question, why you do not have the versions in your vendor_secrets table? what if some ENV is only necessary from some version onward? what if I want to change it but only from version 2.x.y? should you not have the version in there?

Also should we not finally put the semver plugin in postgres?

Reply:
i have mixed fillings about versions. IMO it requires too much maintenance and there are very few cases in which some properties changes from version to version within the table. I would rather use version to define "boundaries". Something like: from the version on ... these are the properties. Then the table will include "ranges":


| service | version | color |
| - | - | - |
| A | 5.0 | red |
| A | 1.1 | blue |
| A | 1.0 | green |


means that service `A` only was green between 1.0 and 1.1, then blue and in 5.0 (latest) it turned red. 
This obviously assumes that:
- service A has published all versions ranges in between (which might not be true)
-

Reply:
well. because of this you cannot use the existing relation to services_meta_data.
Also I can see things like:
s4l-lite:2.1.343 needs ENV XXX_BLAHBLAH
from s4l-lite:3.1.2 needs not XXX_BLAHBLAH but also YYY_BLIHBLAHBLOUH
I agree that in this case, it is probaly harmless to pass the same to 2.1.343.
But what if the use-case is like so:
- SOME_HOST=http://some_url
and from version 3.1.2 it needs to be another protocol such as tcp://blahblahblah ??
In the second one then you are doomed.

Reply:
pair reviewed this changes with @sanderegg @matusdrobuliak66